#pragma once

#include <cstdio>

#include <hip/hip_runtime.h>

#include "legacy/include/flash.h"
#include "legacy/include/intrinsic.h"

namespace flash {
using ::__llvm_fma_f32;
using ::DownCast;
using ::DownCastPairNoPack;
using ::splitkv_upcast_to_f32;
}  // namespace flash

#include "legacy/src/flash_fwd_reduce.h"

namespace gfx93::fwd::dsa_mls {

template<typename Kernel_traits, const bool Tail, typename Params>
void run_dsa_mla_splitkv_reduce(Params& params, hipStream_t stream) {
    static_assert(Kernel_traits::kHeadDimV == 512,
                  "run_dsa_mla_splitkv_reduce only supports hdimv == 512");
    using Element = typename Kernel_traits::Element;
    using SplitkvAccumType = typename Kernel_traits::SplitkvAccumType;

    Flash_fwd_mla_reduce_params reduce_params;
    reduce_params.softmax_lse_ptr = params.softmax_lse_ptr;
    reduce_params.oaccum_ptr = params.oaccum_ptr;
    reduce_params.o_ptr = params.o_ptr;
    reduce_params.cu_seqlens_k = params.cu_seqlens_k;
    reduce_params.num_splits = params.num_splits;
    reduce_params.partition_size = params.partition_size;
    reduce_params.h = params.h;
    reduce_params.ngroups = params.ngroups;
    reduce_params.seqlen_q = params.seqlen_q;
    reduce_params.layout = params.layout;
    reduce_params.topk_length = params.topk_length;
    reduce_params.attn_sink = params.attn_sink;
    reduce_params.extra_topk_length = params.extra_topk_length;
    reduce_params.topk = params.topk;
    reduce_params.extra_topk = params.extra_topk;

    if (params.num_splits > 1) {
        dim3 block(256);
        dim3 grid(params.b * params.h * params.seqlen_q, 4);
        constexpr int MAX_NUM_SPLITS = 64;
        if (params.num_splits > MAX_NUM_SPLITS) {
            printf("\x1b[31mnum_splits %d is larger than limit %d, and thus won't execute the kernel\033[0m\n",
                   params.num_splits, MAX_NUM_SPLITS);
            return;
        }
        if (params.num_splits == 2) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 2, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else if (params.num_splits == 4) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 4, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else if (params.num_splits == 8) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 8, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else if (params.num_splits == 16) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 16, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else if (params.num_splits == 32) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 32, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else if (params.num_splits == 64) {
            ::flash_mla_splitkv_reduce_kernel<SplitkvAccumType, Element, 64, true, Tail, Kernel_traits::kHeadDimV>
                <<<grid, block, 0, stream>>>(reduce_params);
        } else {
            printf("\x1b[31mnum_splits %d is not supported yet, and thus won't execute the kernel\033[0m\n",
                   params.num_splits);
        }
    }
}

}  // namespace gfx93::fwd::dsa_mls
